### ISSUE TITLE
[jvm] Inconsistent prediction for dart booster in java for the same model trained and performing in python. Related issues #3015 #3485 #3540 

### ISSUE THREAD

**Issue by jayantjain93** [dmlc/xgboost#3661]
I am using a dart booster trained in python and loading it in jvm, the prediction seem off. This issue is related to #3015  #3485 #3540 and I tried setting up all the parameters given here, but to no end. I am getting different results in python as suppose to in jvm.

The model and the request file are the same for both the tests as described below.

JAVA code used:
```
public class Model {
    static int batch_size = 50;
    static Booster booster;
    static List<Long> listOfTimers = new LinkedList<Long>();
    public static void main(String[] args) throws Exception{
        HashMap<String, Object> params = new HashMap<String, Object>();
        params.put("numtrees", 30);
        params.put("nthread", 1);
        booster = XGBoost.loadModel("./model_1.model");
        booster.setParams(params);
        File file = new File("./requests.csv");
        Scanner scanner = new Scanner(file);
        int i = 0;
        int size = batch_size*9;
        float[] data = new float[size];
        while (scanner.hasNext()) {
            if(i == batch_size) {
                hitRequests(data, batch_size, 9);
                data = new float[size];
                i = 0;
            }
            String str = scanner.nextLine();
            String [] s = str.split(",");
            for(int j = 0; j < 9; j++) {
                data[i+j] = Float.parseFloat(s[j]);
            }
            i++;

        }
        System.out.println(listOfTimers.toString());
    }

    public static void hitRequests(float [] data, int rows, int cols) throws Exception{
        long time = System.currentTimeMillis();
        DMatrix dMatrix = new DMatrix(data, rows, cols);
        float [][] result = booster.predict(dMatrix, false, 30);
        long end_time = System.currentTimeMillis();
        listOfTimers.add(end_time-time);
        System.out.println("result: " + (getResult(result)));
        System.out.println("timer: " + (end_time-time));
    }
   
}
```
results that i get:
```
result: [0.61936563], [0.5984782], [0.32919276], [0.5649111], [0.47643474], [0.3365553], [2.3018966], [1.9958245], [1.9958245], [1.9958245], [1.9958245], [1.9958245], [1.9958245], [1.9958245], [1.9958245], [1.9958245], [1.9958245], [1.9958245], [1.9958245], [1.9958245], [1.9958245], [1.9958245], [1.9958245], [1.9958245], [1.9958245], [1.9958245], [1.9958245], [1.9958245], [1.9958245], [1.9958245], [1.9958245], [1.9958245], [1.9958245], [1.9958245], [1.9958245], [1.9958245], [1.9958245], [1.9958245], [1.9958245], [1.9958245], [1.9958245], [1.9958245], [1.9958245], [1.9958245], [1.9958245], [1.9958245], [1.9958245], [1.9958245], [1.9958245], [1.9958245]
```

Python code and results added below, when parsing the same model and request file:
python code:
```
from pathlib import Path
import xgboost as xg
import pandas as pd
import time

def request_hit(data):
    t1 = time.time()
    dmat = xg.DMatrix(data=data)
    pred = model.predict(data=dmat)
    t2 = time.time()
    print(str(pred))
    print(str(int((t2-t1) *1000)))

def parseFeatureFile(file_name: Path):
    with file_name.open('r') as file:
        feature = []
        for line in file:
            vector = []
            line = line.rstrip("\n")
            stringFeatures = line.split(",")
            for i in stringFeatures:
                vector.append(float(i))
            feature.append(vector)
        return feature


if __name__ == '__main__':
    model_file = "../../Model/model_1.model"

    model = xg.Booster()
    model.load_model(model_file)
    model.set_param("numtrees", 30)
    model.set_param("nthread", 1)

    request_file = parseFeatureFile(Path("../../Model/requests.csv"))
    i = 0
    data = []
    for row in request_file:
        if i == 50:
            request_hit(data=data)
            i = 0
            data = []
            time.sleep(1)
        data.append(row)
        # print(str(row))
        i += 1
```
python results:
```
[ 1.23176956  4.36962748  0.36619836  0.30802074  0.61152518  0.61152518
  0.69353884  0.8411895   0.63114357  0.63114357  0.36619836  0.63114357
  1.98527622  0.25395423  0.25061363  2.02189493  0.40367284  0.49325848
  0.61167997  0.61167997  0.4122296   4.18055201  0.35991892  0.59111851
  0.69077677  0.7807771   0.40367284  0.43420082  1.82767808  2.02189493
  0.59111851  0.75833762  0.31537506  2.37626696  0.3091521   0.76201147
  0.5229069   0.41403303  0.65020835  1.58646369  0.45272243  1.63795483
  0.77941149  0.17258394  0.17258394  0.29119542  0.5229069   1.83758581
  0.58843833  0.93746954]
```

sample request file:
```
18411.0, 12.0, 1.0, 0.0, 12.0, 0.0, 0.0, 4.0, 6.0
11198.0, 12.0, 1.0, 0.0, 80.0, 0.0, 0.0, 4.0, 6.0
23724.0, 7.0, 1.0, 0.0, 15.0, 24.0, 0.0, 4.0, 6.0
16737.0, 3.0, 1.0, 0.0, 59.0, 5.0, 0.0, 4.0, 6.0
```

Not sure with the there is bug in the java implementation or if I have done something wrong.

**Comment by hcho3** [comment]
You should add the parameter `ntree_limit` to get deterministic prediction:
```python
preds = bst.predict(dtest, ntree_limit=num_round)
```
(Java will probably behave the same)

Without `ntree_limit`, the `predict()` method will perform dropout. This behavior is documented in https://xgboost.readthedocs.io/en/latest/tutorials/dart.html.

**Comment by hcho3** [comment]
Yes, `treeLimit` is the one.

**Comment by jayantjain93** [comment]
@hcho3 Sorry for the late reply. 
That didn't help either, I had tried it in the original test.
I think there might be a bug.

**Comment by hcho3** [comment]
@jayantjain93 Did you specify `ntree_limit` on the Python side too?

**Comment by jayantjain93** [comment]
@hcho3 , Yes I had specified it in python as well.
By not specifying it was giving inconsistent result in python.
